modules/measure/src/main/java/com/opengamma/strata/measure/fxopt/FxVanillaOptionTradeCalculationFunction.java

```diff
@@ -0,0 +1,174 @@
+/**
+ * Copyright (C) 2016 - present by OpenGamma Inc. and the OpenGamma group of companies
+ *
+ * Please see distribution for license.
+ */
+package com.opengamma.strata.measure.fxopt;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.opengamma.strata.basics.ReferenceData;
+import com.opengamma.strata.basics.currency.Currency;
+import com.opengamma.strata.basics.currency.CurrencyPair;
+import com.opengamma.strata.calc.Measure;
+import com.opengamma.strata.calc.runner.CalculationFunction;
+import com.opengamma.strata.calc.runner.CalculationParameters;
+import com.opengamma.strata.calc.runner.FunctionRequirements;
+import com.opengamma.strata.collect.result.FailureReason;
+import com.opengamma.strata.collect.result.Result;
+import com.opengamma.strata.data.scenario.ScenarioMarketData;
+import com.opengamma.strata.measure.Measures;
+import com.opengamma.strata.measure.rate.RatesMarketDataLookup;
+import com.opengamma.strata.measure.rate.RatesScenarioMarketData;
+import com.opengamma.strata.product.fxopt.FxVanillaOption;
+import com.opengamma.strata.product.fxopt.FxVanillaOptionTrade;
+import com.opengamma.strata.product.fxopt.ResolvedFxVanillaOptionTrade;
+
+/**
+ * Perform calculations on an FX vanilla option trade for each of a set of scenarios.
+ * <p>
+ * This uses Black FX option volatilities, which must be specified using {@link FxOptionMarketDataLookup}.
+ * An instance of {@link RatesMarketDataLookup} must also be specified.
+ * <p>
+ * Two pricing methods are available, 'Black' and 'VannaVolga'.
+ * By default, 'Black' will be used. To control the method, pass an instance of
+ * {@link FxVanillaOptionMethod} in the calculation parameters.
+ * <p>
+ * The supported built-in measures are:
+ * <ul>
+ *   <li>{@linkplain Measures#PRESENT_VALUE Present value}
+ *   <li>{@linkplain Measures#PV01_CALIBRATED_SUM PV01 calibrated sum on rate curves}
+ *   <li>{@linkplain Measures#PV01_CALIBRATED_BUCKETED PV01 calibrated bucketed on rate curves}
+ *   <li>{@linkplain Measures#PV01_MARKET_QUOTE_SUM PV01 market quote sum on rate curves}
+ *   <li>{@linkplain Measures#PV01_MARKET_QUOTE_BUCKETED PV01 market quote bucketed on rate curves}
+ *   <li>{@linkplain Measures#CURRENCY_EXPOSURE Currency exposure}
+ *   <li>{@linkplain Measures#CURRENT_CASH Current cash}
+ *   <li>{@linkplain Measures#RESOLVED_TARGET Resolved trade}
+ * </ul>
+ * <p>
+ * The "natural" currency is the base currency of the market convention pair of the optional FX.
```

Comment:
'the optional FX' confused me - I think best to refer to this as something like 'the base currency of the underlying currency pair'.